### PR TITLE
Make dedicated fiber-photometry session json generation

### LINKED ISSUE #250: Update fiber-photometry session json generation
**Is your feature request related to a problem? Please describe.**
From Kenta:

At the moment, fiber photometry sessions can be subdivided into:
1.Dynamic Foraging + FP
2.Pavlovian Conditioning + FP
3.Optogenetic stimulation + FP
 
For 1, [Dynamic Foraging experimental code](https://github.com/AllenNeuralDynamics/dynamic-foraging-task) is generating sessions.json upon data collection by NOT using metadata mapper (and thus getting obsolete over time). This was designed and built by a team effort with Xinxin, Alex, and Han as the primary contributors. 
 
For 2, I wrote a dirty snippet to generate session json offline (=not during experiments) for Pavlovian Conditioning. So again, not utilizing mapper function.
 
For 3, there is a super old metadata mapper written by Mekhla in 2023. It got obsolete, and thus, I wrote a dirty snippet to generate session json offline too.
 
(Addressing 1 would be more elaborate effort as the DF GUI software codebase is huge, many people are involved, and downstream pipeline is also getting larger now.)

**Describe the solution you'd like**
We want to attempt to generate metadata for the fiber rigs and migrate it to aind-meta-data mapper to prevent things from getting obsolete. This will make it easier to propogate breaking changes from the data schema

**Additional context**
An additional step would be to remove the metadata generation here: https://github.com/AllenNeuralDynamics/dynamic-foraging-task/blob/main/src/foraging_gui/GenerateMetadata.py. We can submit a PR in this repo

### PR DESCRIPTION
**Pull request by dougollerenshaw**
# Simplify and Improve Fiber Photometry ETL Implementation

## Motivation
The previous implementation was very specific to one experimental paradigm where fiber photometry was combined with optogenetic stimulation. We want a general fiber photometry session generator which can be paired with any other experimental platform. This PR accomplishes that by creating a more modular and focused implementation.

### Naming and Organization
- Created a dedicated `utils.py` in the `fip` module to house helper functions, improving code organization and maintainability
- Moved time extraction and file handling logic to `utils.py`, keeping the ETL implementation clean and focused

### Core Functionality
- Removed all references to optogenetic stimulation experiment in `fip/session.py`
- Implemented automatic session time extraction through:
  - Start time: Parsed from data file names (format: `FIP_DataG_YYYY-MM-DDThh_mm_ss.csv`)
  - End time: Calculated from timestamps in CSV data files
  - Proper timezone handling for consistent UTC conversion
- Simplified the ETL interface to focus solely on fiber photometry data processing

### Documentation
- Added comprehensive README ([direct link](https://github.com/AllenNeuralDynamics/aind-metadata-mapper/blob/feat-250-update-fiber-metadata/src/aind_metadata_mapper/fip/README.md)) with:
  - Clear usage examples
  - Command-line interface documentation
  - Explanation of automatic time extraction
  - Description of expected file formats and naming conventions

### Testing
- Maintained 100% test coverage for FIP module
- Added specific tests for:
  - Time extraction utilities
  - File handling functions
  - Timezone conversion edge cases
  - Error handling scenarios

## Impact
These changes make the fiber photometry ETL more:
- Modular: Can be easily integrated with other experimental platforms
- Maintainable: Clear separation of concerns between ETL and utilities
- Robust: Better error handling and time zone management
- User-friendly: Improved documentation and examples

## Important note on standards/file conventions
This PR makes a number of implicit assumptions about the file naming conventions and the file structures as outlined in the v0.2.2 of the file standards doc [here](https://github.com/AllenNeuralDynamics/aind-file-standards/blob/313f1afc1eef7512335c2c494c952268dad9bd28/file_formats/fip.md). If those naming conventions change, this mapper will cease to work. Ideally we would have a more structured way of defining and documenting these conventions to avoid confusion.

Closes #250

### CODE REVIEW

File: .github/workflows/run_main_tests.yml [hunk 1]
```diff
@@ -22,9 +22,15 @@ jobs:
       run: | 
         python -m pip install -e .[dev]
     - name: Run linter checks
-      run: flake8 . && interrogate --verbose .
+      run: |
+        echo "=== Running style checks with flake8 ==="
```
Comment (arielleleon) [inline .github/workflows/run_main_tests.yml:26]: Can you ask Jon if he can add these prints to the template?

**Review by arielleleon — APPROVED** [review]

File: .github/workflows/run_main_tests.yml (hunk 1 above)
Comment (dougollerenshaw) [reply .github/workflows/run_main_tests.yml:26]: Thanks @arielleleon for the review! Agreed that these print statements would be useful in the template and other repos. I was motivated to add this because I was confused about the coverage report generated by the `interrogate --verbose .` I interpreted it as a test coverage report, when it was actually a docstring coverage report. So I spent quite a long time (more than I want to admit!) trying to understand why I had 100% test coverage locally but couldn't get past this check. 

@jtyoung84 , tagging you for visibility. I'm about to request review from you also.

**Review by arjunsridhar12345 — APPROVED** [review]

File: src/aind_metadata_mapper/fib/models.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+    experimenter_full_name: List[str]
+    session_start_time: Optional[datetime] = None
+    session_end_time: Optional[datetime] = None
+    subject_id: str
+    rig_id: str
+    mouse_platform_name: str
+    active_mouse_platform: bool
+    data_streams: List[dict]
+    session_type: str = "Fiber_Photometry"
```
[the remark this answers is not in the document]
Comment (dougollerenshaw) [reply src/aind_metadata_mapper/fib/models.py]: Good point. Yes, it probably should. See https://github.com/AllenNeuralDynamics/aind-metadata-mapper/pull/251/commits/de7a1d9a333f8d2ed7805d51c4f6b967228e7a4c. 

One complication, though, is that when this is joined with the corresponding behavior metadata, we'll have to figure out how to reconcile the session types

**Comment by dougollerenshaw** [comment]
Important note here. The exposure time in the data streams is wrong. See https://github.com/AllenNeuralDynamics/aind-data-migration-scripts/issues/355. I'll need to be sure to populate with correct values from @hagikent before merging.

File: src/aind_metadata_mapper/fib/utils.py [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+from typing import Union, Optional
+from pathlib import Path
+from tzlocal import get_localzone
+import logging
+import pandas as pd
+import re
+import json
+
+
+def convert_ms_since_midnight_to_datetime(
```
Comment (jtyoung84) [inline src/aind_metadata_mapper/fib/utils.py:19]: Timezone handling is notoriously tricky. There are many edge cases that need to be tested. In general, it's best to convert everything to timestamps or UTC. For example, python's datetime object doesn't handle daylight savings time very well. Assuming that the ms_since_midnight are the actual milliseconds elapsed since midnight of the local time, then converting the local time time first and then applying the time delta circumvents that edge case:
```python
from datetime import datetime, timedelta, time
from zoneinfo import ZoneInfo
from aind_metadata_mapper.fib.utils import convert_ms_since_midnight_to_datetime

def convert_ms(ms_since_midnight, base_date):
    base_date_midnight = datetime.combine(base_date.date(), datetime.min.time())
    base_date_midnight_utc = base_date_midnight.astimezone(ZoneInfo("UTC"))
    delta = timedelta(milliseconds=ms_since_midnight)
    return base_date_midnight_utc + delta

base_date = datetime(2025, 3, 9, 0, 0, 0, tzinfo=ZoneInfo(key='America/Los_Angeles'))

# These output the same datetime object even though the millis are different
convert_ms_since_midnight_to_datetime(1000*60*60*2.5, base_date)
# datetime.datetime(2025, 3, 9, 10, 30, tzinfo=zoneinfo.ZoneInfo(key='UTC'))

convert_ms_since_midnight_to_datetime(1000*60*60*3.5, base_date)
# datetime.datetime(2025, 3, 9, 10, 30, tzinfo=zoneinfo.ZoneInfo(key='UTC'))

# These output different objects
convert_ms(1000*60*60*2.5, base_date)
# datetime.datetime(2025, 3, 9, 10, 30, tzinfo=zoneinfo.ZoneInfo(key='UTC'))

convert_ms(1000*60*60*3.5, base_date)
# datetime.datetime(2025, 3, 9, 11, 30, tzinfo=zoneinfo.ZoneInfo(key='UTC'))
```

**Review by jtyoung84 — COMMENTED** [review]
I left one note about the time zone handling that may need to be resolved. I'll leave a few comments without requesting changes:

- We've been using NumPy's docstring format for our docstring formats
- We discourage writing temp files during unit testing. We usually mock the write and compare things in-memory
- We encourage large PRs to be broken up into ~500 line changes if possible

**Review by jtyoung84 — CHANGES_REQUESTED** [review]
The `convert_ms_since_midnight_to_datetime` may need to be updated to handle edge cases.

File: src/aind_metadata_mapper/fib/session.py [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+                    status_code=500,
+                    message=f"Error writing to {output_path}: {str(e)}",
+                    data=None,
+                )
+
+        # Otherwise use the default write_standard_file method
+        return self._load(transformed, self.job_settings.output_directory)
+
+
+def update_job_settings_with_times(
```
Comment (jtyoung84) [inline src/aind_metadata_mapper/fib/session.py:257]: It seems like this method should either be added to the utils package or be added as a class method.

File: src/aind_metadata_mapper/fib/session.py [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+
+        Raises:
+            ValidationError: If the provided settings fail schema validation
+            JSONDecodeError: If job_settings is a string but not valid JSON
+        """
+        if isinstance(job_settings, str):
+            job_settings = JobSettings(**json.loads(job_settings))
+        super().__init__(job_settings)
+
+    def _extract(self) -> JobSettings:
```
Comment (jtyoung84) [inline src/aind_metadata_mapper/fib/session.py:67]: The extract methods usually parse the input sources into intermediate data models that can be transformed into the output Session object.

File: src/aind_metadata_mapper/fib/session.py [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+                )
+            else:
+                logging.warning(
+                    "No data_directory provided and "
+                    "no session_start_time specified"
+                )
+
+        return settings
+
+    def _create_stream(
```
Comment (jtyoung84) [inline src/aind_metadata_mapper/fib/session.py:105]: This method could be decorated as a `@staticmethod` and the `self` arg can be removed. Alternatively, the settings could potentially be accessed directly via `self.job_settings`.

File: src/aind_metadata_mapper/fib/utils.py (hunk 3 above)
Comment (dougollerenshaw) [reply src/aind_metadata_mapper/fib/utils.py:19]: Good point. I updated the function as you suggested and also added specific test cases for March 9, 2025 (the day we switched to DST in 2025). See https://github.com/AllenNeuralDynamics/aind-metadata-mapper/pull/251/commits/6ba5ba9dc9b55e32ac99522ee4019ab81468b811

**Review by hagikent — APPROVED** [review]
main functions worked

File: src/aind_metadata_mapper/fib/session.py (hunk 4 above)
Comment (dougollerenshaw) [reply src/aind_metadata_mapper/fib/session.py]: I removed this function entirely.

File: src/aind_metadata_mapper/fib/session.py (hunk 5 above)
Comment (dougollerenshaw) [reply src/aind_metadata_mapper/fib/session.py]: I made some heavy modifications to session.py. Let me know if that satisfies you.

File: src/aind_metadata_mapper/fib/session.py (hunk 6 above)
Comment (dougollerenshaw) [reply src/aind_metadata_mapper/fib/session.py]: Removed entirely.

**Comment by dougollerenshaw** [comment]
@jtyoung84:

> I left one note about the time zone handling that may need to be resolved. I'll leave a few comments without requesting changes:

Timezone fix implemented.

> 
> We've been using NumPy's docstring format for our docstring formats

fixed

> We discourage writing temp files during unit testing. We usually mock the write and compare things in-memory

fixed. using mock now

> We encourage large PRs to be broken up into ~500 line changes if possible

One reason for the big diff was that I had renamed the module from "fip" to "fib" to be consistent. So there were a lot of full file deletions/additions. But I reverted those changes to keep the diff more manageable. I'm still really bothered by this naming convention mismatch, though, and think it should be addressed in a followup PR. It's complicated because the scientific team uses "FIP" but the SciComp standard seems to be "FIB" (e.g. we require that fiber data be in a folder named "fib"). So we should probably work to get everyone to agree on one or the other and change it everywhere to match!

**Review by jtyoung84 — APPROVED** [review]
